### ISSUE TITLE
NCCL hang in PyTorch Distributed Data Parallel for Mixed Precision Training

### ISSUE THREAD

**Issue by 8enmann** [pytorch/pytorch#20125]
## 🐛 Bug

On an AWS p3.16xlarge machine, I'm using the mixed precision training from https://github.com/NVIDIA/Megatron-LM checked in at https://github.com/cybertronai/transformer-xl
When I run with [adaptive softmax](https://github.com/cybertronai/transformer-xl/blob/master/mem_transformer.py#L477) with 8 GPUs and [apex mixed precision](https://github.com/cybertronai/transformer-xl/blob/master/fp16_opt.py), NCCL hangs. When I remove `--adaptive`, it works as expected (after reducing batch size to avoid OOM). If I remove `--fp16` it also works fine, so it's definitely a mixed precision issue.

Sorry if this is a dup of #11672 but I haven't been able to figure it out.

## To Reproduce

On a p3.16xlarge
```sh
git clone https://github.com/cybertronai/transformer-xl
cd transformer-xl
pip install -r requirements.txt
# Download dataset to ./data/wikitext-103/ then
NCCL_DEBUG_FILE=log.%h.%p NCCL_DEBUG_SUBSYS=COLL NCCL_DEBUG=INFO NCCL_MIN_NRINGS=16 NCCL_MAX_NRINGS=16  python -m tor
ch.distributed.launch --nproc_per_node=8 --nnodes=1 --node_rank=0 --master_addr=172.31.20.228 --master_port=6016 train.py --seed 1111 --data data/wikitext-103
 --dataset wt103 --adaptive --log_interval 100 --eval_interval 500 --max_tokens 1500000000 --logdir /ncluster/runs/ben-no-bpe.13 --distributed --lr 0.0005 --b
atch_size 4 --eta_min 5e-05 --n_layer 18 --d_model 1024 --n_head 16 --d_head 64 --d_inner 4096 --dropout 0.2 --dropatt 0.2 --optim lamb --warmup_tokens 300000
0 --tgt_len 128 --mem_len 128 --eval_tgt_len 128 --fp16 --dynamic_loss_scale  --init_std 0.005 --div_val 4 
```

In the [logs](https://github.com/NVIDIA/nccl/files/3144052/combined.txt) I found this:
```
ip-172-31-20-228:22326:22707 [3] NCCL INFO AllReduce: opCount 5 sendbuff 0x7fcf6b400000 recvbuff 0x7fcf6b400000 count 1322459 datatype 6 op 0 root 0 comm 0x7f
cfc403a800 [nranks=8] stream 0x563efa445460
ip-172-31-20-228:22327:22327 [4] NCCL INFO AllReduce: opCount 5 sendbuff 0x7fb11f71c200 recvbuff 0x7fb11f71c200 count 1 datatype 4 op 0 root 0 comm 0x7fb09c03
a800 [nranks=8] stream 0x55c779123600
```
How do I figure out where this is coming from?  Is this something my code should account for or is it a bug in DDP or apex?

## Expected behavior
Doesn't hang silently.

## Environment

Please copy and paste the output from our
[environment collection script](https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py)
(or fill out the checklist below manually).

```
PyTorch version: 1.0.1
Is debug build: No
CUDA used to build PyTorch: 10.0.130

OS: Ubuntu 16.04.6 LTS
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.11) 5.4.0 20160609
CMake version: version 3.12.2

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 10.0.130
GPU models and configuration:
GPU 0: Tesla V100-SXM2-16GB
GPU 1: Tesla V100-SXM2-16GB
GPU 2: Tesla V100-SXM2-16GB
GPU 3: Tesla V100-SXM2-16GB
GPU 4: Tesla V100-SXM2-16GB
GPU 5: Tesla V100-SXM2-16GB
GPU 6: Tesla V100-SXM2-16GB
GPU 7: Tesla V100-SXM2-16GB

Nvidia driver version: 410.104
cuDNN version: Could not collect

Versions of relevant libraries:
[pip3] numpy==1.15.4
[conda] blas                      1.0                         mkl
[conda] cuda100                   1.0                           0    pytorch
[conda] mkl                       2019.3                      199
[conda] mkl-include               2019.3                      199
[conda] mkl-service               1.1.2            py36he904b0f_5
[conda] mkl_fft                   1.0.10           py36ha843d7b_0
[conda] mkl_random                1.0.2            py36hd81dba3_0
[conda] pytorch                   1.0.1           py3.6_cuda10.0.130_cudnn7.4.2_0    pytorch
[conda] pytorch-lamb              0.0.0                     <pip>
[conda] pytorch-pretrained-bert   0.6.2                     <pip>
[conda] torchvision               0.2.1                      py_2    pytorch
```

**Comment by 8enmann** [comment]
This appears to have been fixed in 1.1, will reopen if it's broken after upgrade.

**Comment by 8enmann** [comment]
Nope, still broken
```
PyTorch version: 1.1.0
Is debug build: No
CUDA used to build PyTorch: 9.0.176

OS: Ubuntu 16.04.6 LTS
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.11) 5.4.0 20160609
CMake version: version 3.12.2

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 10.0.130
GPU models and configuration:
GPU 0: Tesla V100-SXM2-16GB
GPU 1: Tesla V100-SXM2-16GB
GPU 2: Tesla V100-SXM2-16GB
GPU 3: Tesla V100-SXM2-16GB
GPU 4: Tesla V100-SXM2-16GB
GPU 5: Tesla V100-SXM2-16GB
GPU 6: Tesla V100-SXM2-16GB
GPU 7: Tesla V100-SXM2-16GB

Nvidia driver version: 410.104
cuDNN version: Could not collect

Versions of relevant libraries:
[pip3] numpy==1.15.4
[conda] blas                      1.0                         mkl
[conda] cuda100                   1.0                           0    pytorch
[conda] mkl                       2019.3                      199
[conda] mkl-include               2019.3                      199
[conda] mkl-service               1.1.2            py36he904b0f_5
[conda] mkl_fft                   1.0.10           py36ha843d7b_0
[conda] mkl_random                1.0.2            py36hd81dba3_0
[conda] pytorch                   1.0.1           py3.6_cuda10.0.130_cudnn7.4.2_0    pytorch
[conda] pytorch-lamb              0.0.0                     <pip>
[conda] pytorch-pretrained-bert   0.6.2                     <pip>
[conda] torch                     1.1.0                     <pip>
[conda] torchvision               0.2.1                      py_2    pytorch
```

**Comment by 8enmann** [comment]
I ran with n_procs=1 and suddenly a magic new error message appeared telling me to try adding `find_unused_parameters=True` to `DistributedDataParallel`. When I did that, it worked! Thank you for good error messages!!!
